rmw_connextdds_common/include/rmw_connextdds/rmw_waitset_std.hpp

```diff
@@ -333,8 +350,55 @@ class RMW_Connext_StatusCondition : public RMW_Connext_Condition
   virtual bool
   has_status(const rmw_event_type_t event_type) = 0;
 
+  void
+  notify_new_event(rmw_event_type_t event_type)
+  {
+    std::unique_lock<std::mutex> lock_mutex(new_event_mutex_);
+    if (new_event_cb_[event_type]) {
+      new_event_cb_[event_type](user_data_[event_type], 1);
+    } else {
+      unread_events_count_[event_type]++;
+    }
+  }
+
+  void
+  set_new_event_callback(
+    rmw_event_type_t event_type,
+    rmw_event_callback_t callback,
+    const void * user_data)
+  {
+    std::unique_lock<std::mutex> lock_mutex(new_event_mutex_);
+
+    if (callback) {
+      // Push events arrived before setting the executor's callback
+      if (unread_events_count_[event_type] > 0) {
+        callback(user_data, unread_events_count_[event_type]);
+        unread_events_count_[event_type] = 0;
+      }
+      user_data_[event_type] = user_data;
+      new_event_cb_[event_type] = callback;
+    } else {
+      user_data_[event_type] = nullptr;
+      new_event_cb_[event_type] = nullptr;
+    }
+  }
+
+  void
+  on_inconsistent_topic(const struct DDS_InconsistentTopicStatus * status);
+
+  void
+  update_status_inconsistent_topic(const struct DDS_InconsistentTopicStatus * status);
+
 protected:
   DDS_StatusCondition * scond;
+  std::mutex new_event_mutex_;
+  rmw_event_callback_t new_event_cb_[RMW_EVENT_INVALID] = {};
+  const void * user_data_[RMW_EVENT_INVALID] = {};
+  uint64_t unread_events_count_[RMW_EVENT_INVALID] = {0};
+
+  bool triggered_inconsistent_topic{false};
+
+  struct DDS_InconsistentTopicStatus status_inconsistent_topic;
```

Comment:
There changes obviously break the ABI compatibility, are we allowed to do that in this particular time?

Reply:
> There changes obviously break the ABI compatibility, are we allowed to do that in this particular time?

It's true that this breaks ABI.  In this case, I approved it anyway because the ABI is between `rmw_connextdds_common` and `rmw_connextdds`; it seems doubtful someone is relying on that.  Also, for some other RMWs we have broken ABI in this way before if it was "internal" to the RMW.

But it is a conversation worth having; should we allow this kind of ABI breakage in stable distributions?

Reply:
i think that is okay in this case to break the ABI, i do not expect there could be someone relying on rmw implementation APIs...

> But it is a conversation worth having; should we allow this kind of ABI breakage in stable distributions?

IMO, i think rmw implementation can break the internal ABI in stable distributions. as long as RMW interface is compatible, there should be no problem? agree that this is worth to bring up for the discussion.

Reply:
Would you be ok with this change then? I prefer to be sure than to revert.

As I see it, this would only break compatibility from programs that directly call the Connext RMW API, which would be defeating the purpose of the RMW in the first place. The ABI of the `rmw_connextdds` and `rmw` packages is still stable, so it shouldn't be an issue.

Reply:
> Would you be ok with this change then? I prefer to be sure than to revert.

Let's hold off until next week; we'll have a discussion in the ROS PMC meeting about it.